Modules/_lsprof.c

```diff
@@ -780,34 +780,47 @@ _lsprof_Profiler_enable_impl(ProfilerObject *self, int subcalls,
         return NULL;
     }
 
-    if (PyObject_CallMethod(monitoring, "use_tool_id", "is", self->tool_id, "cProfile") == NULL) {
+    PyObject *check = PyObject_CallMethod(monitoring,
+                                          "use_tool_id", "is",
+                                          self->tool_id, "cProfile");
+    if (check == NULL) {
+        PyErr_Clear();
         PyErr_Format(PyExc_ValueError, "Another profiling tool is already active");
-        Py_DECREF(monitoring);
-        return NULL;
+        goto error;
     }
+    Py_DECREF(check);
 
     for (int i = 0; callback_table[i].callback_method; i++) {
+        int event = (1 << callback_table[i].event);
         PyObject* callback = PyObject_GetAttrString((PyObject*)self, callback_table[i].callback_method);
         if (!callback) {
-            Py_DECREF(monitoring);
-            return NULL;
+            goto error;
+        }
+        PyObject *register_result = PyObject_CallMethod(monitoring, "register_callback",
+                                                        "iiO", self->tool_id,
+                                                        event, callback);
+        Py_DECREF(callback);  // register_result is callback
+        if (register_result == NULL) {
+            goto error;
         }
-        Py_XDECREF(PyObject_CallMethod(monitoring, "register_callback", "iiO", self->tool_id,
-                                       (1 << callback_table[i].event),
-                                       callback));
-        Py_DECREF(callback);
-        all_events |= (1 << callback_table[i].event);
```

Comment:
You still need to decref `register_result`

Reply:
Indeed. See:
- https://github.com/python/cpython/blob/78842e4a98994a218a93992a2a1e3ca3eaa28e79/Python/instrumentation.c#L1441
- https://github.com/python/cpython/blob/78842e4a98994a218a93992a2a1e3ca3eaa28e79/Python/instrumentation.c#L2266

We pass `callback` to `_PyMonitoring_RegisterCallback` which increfs it once again.